buildSrc/src/main/groovy/org/elasticsearch/gradle/NoticeTask.groovy

```diff
@@ -50,46 +55,122 @@ public class NoticeTask extends DefaultTask {
-                        throw new RuntimeException("Two different notices exist for dependency '" +
-                                name + "': " + prevFile + " and " + file)
-                    }
-                } else {
-                    seen.put(name, file)
+        noticeFiles.each { File file ->
+            String name = file.name.replaceFirst(/-NOTICE\.txt$"/, "")
+            if (seen.containsKey(name)) {
+                File prevFile = seen.get(name)
+                if (prevFile.text != file.text) {
+                    throw new RuntimeException("Two different notices exist for dependency '" +
+                            name + "': " + prevFile + " and " + file)
                 }
+            } else {
+                seen.put(name, file)
             }
         }
+
+        // Add all LICENSE and NOTICE files in licenses directory
         for (Map.Entry<String, File> entry : seen.entrySet()) {
             String name = entry.getKey()
             File file = entry.getValue()
             appendFile(file, name, 'NOTICE', output)
             appendFile(new File(file.parentFile, "${name}-LICENSE.txt"), name, 'LICENSE', output)
         }
+
+        // Find any source files with "@notice" annotated license header
+        for (File sourceFile : sources.files) {
+            boolean isPackageInfo = sourceFile.name == 'package-info.java'
+            boolean foundNotice = false
+            boolean inNotice = false
+            StringBuilder header = new StringBuilder()
+            String packageDeclaration
+
+            for (String line : sourceFile.readLines()) {
+                if (isPackageInfo && packageDeclaration == null && line.startsWith('package')) {
+                    packageDeclaration = line
+                }
+
+                if (foundNotice == false) {
+                    foundNotice = line.contains('@notice')
+                    inNotice = true
+                } else {
+                    if (line.contains('*/')) {
+                        inNotice = false
+
+                        if (!isPackageInfo) {
+                            break
+                        }
+                    } else if (inNotice) {
+                        header.append(line.stripMargin('*'))
+                        header.append('\n')
+                    }
+                }
+            }
+
+            if (foundNotice) {
+                appendText(header.toString(), isPackageInfo ? packageDeclaration : sourceFile.name, '', output)
+            }
+        }
         outputFile.setText(output.toString(), 'UTF-8')
     }
 
+    @InputFiles
+    @Optional
+    FileCollection getNoticeFiles() {
+        FileTree tree
+        licensesDirs.each { dir ->
+            if (tree == null) {
+                tree = project.fileTree(dir)
+            } else {
+                tree += project.fileTree(dir)
+            }
+        }
+
+        return tree?.matching { include '**/*-NOTICE.txt' }
```

Comment:
Shouldn't this always be not-null since we set it above?

Reply:
We only set it if `licenseDirs` is non-empty. Not all subprojects have a license dir, so this needs to be optional.